### PR TITLE
Cross-Account e2e Tests + Conflicted Gateway Status

### PR DESCRIPTION
**Pull request by xWink**
**What type of PR is this?**

feature + tests

**Which issue does this PR fix**:
#567 

**What does this PR do / Why do we need it**:
- Updated Gateway Controller to appropriately set the Programmed status of Gateways as `Conflicted` when there is a name conflict
    - Previously, the status would be left as `Pending`
- Added new cross-account e2e test suite, named `sharing`
    - Not executed as part of normal e2e-test suite due to different environment expectations and testing goals
- Added new environment variable `SECONDARY_ACCOUNT_TEST_ROLE_ARN`, which is used by the new test suite to assume the secondary account, which currently just shares resources to the primary account
    - In the future, the secondary account may be expected to have its own cluster, which would enable multi-cluster e2e testing in the new suite
- Added new makefile target `sharing-test`, which runs the new suite
- Added docs explaining how to use the `sharing` suite

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:
N/A

**Testing done on this change**:

```
Ran 3 of 3 Specs in 118.312 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (118.31s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/sharing       119.038s
```

**Automation added to e2e**:

`sharing` suite

**Will this PR introduce any new dependencies?**:

No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
No

**Does this PR introduce any user-facing change?**:

No

```release-note
Added cross-account e2e test suite
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -121,7 +121,21 @@ e2e-test: ## Run e2e tests against cluster pointed to by ~/.kube/config
 		-count 1 \
 		-timeout 90m \
 		-v \
-		./suites/... \
+		./suites/integration/... \
+		--ginkgo.focus="${FOCUS}" \
+		--ginkgo.timeout=90m \
+		--ginkgo.v
```
Comment (mikhail-aws) [inline Makefile:127]: are these same flags from above?
		-count 1 \
		-timeout 90m \

File: Makefile [hunk 2]
```diff
@@ -121,7 +121,21 @@ e2e-test: ## Run e2e tests against cluster pointed to by ~/.kube/config
...
+	LOG_LEVEL=debug
+	cd test && go test \
+		-p 1 \
+		-count 1 \
+		-timeout 90m \
+		-v \
+		./suites/sharing/... \
 		--ginkgo.focus="${FOCUS}" \
 		--ginkgo.timeout=90m \
 		--ginkgo.v
```
Comment (mikhail-aws) [inline Makefile:141]: please dont copy paste code

File: pkg/config/controller_config.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ const (
 	CLUSTER_VPC_ID                  = "CLUSTER_VPC_ID"
 	CLUSTER_NAME                    = "CLUSTER_NAME"
 	DEFAULT_SERVICE_NETWORK         = "DEFAULT_SERVICE_NETWORK"
+	SECONDARY_ACCOUNT_TEST_ROLE_ARN = "SECONDARY_ACCOUNT_TEST_ROLE_ARN"
```
Comment (mikhail-aws) [inline pkg/config/controller_config.go:26]: should be in test package not source code

File: pkg/controllers/gateway_controller.go [hunk 4]
```diff
@@ -224,18 +224,35 @@ func (r *gatewayReconciler) reconcileUpsert(ctx context.Context, gw *gwv1beta1.G
...
+		}
+		return nil
+	}
+
+	if len(foundSnArns) > 1 {
+		if err = r.updateGatewayProgrammedStatus(ctx, gw, core.GatewayReasonConflicted, "Multiple service networks with the same name were found. Either ensure only one service network with the same name is visible in the AWS account or use the desired service network's id as the Gateway name."); err != nil {
+			return lattice_runtime.NewRetryError()
+		}
+		return nil
+	}
```
Comment (mikhail-aws) [inline pkg/controllers/gateway_controller.go:253]: - status updates for Accepted and Programmed should be in single API call, update both statuses and then call k8s-api
- group validation logic into own method
- UpdateGWListenerStatus already doing some job, need to refactor. Structure should be UpdateStatus -> UpdateAccepted, UpdateProgrammed -> Call K8s API with all statuses populated

File: pkg/model/core/status_condition.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package core
+
+import gwv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+const GatewayReasonConflicted gwv1.GatewayConditionReason = "Conflicted"
```
Comment (mikhail-aws) [inline pkg/model/core/status_condition.go:5]: should be in controller's package next to GW controller, status update not related to models

File: pkg/utils/common.go [hunk 6]
```diff
@@ -85,6 +86,14 @@ func TargetRefToLatticeResourceName(
...
 }
 
+func RandomString(length int) string {
+	randomSuffix := make([]rune, length)
+	for i := range randomSuffix {
+		randomSuffix[i] = rune(rand.Intn(26) + 'a')
+	}
+	return string(randomSuffix)
+}
+
```
Comment (mikhail-aws) [inline pkg/utils/common.go:96]: not generic random string, and should not be used in real code, please move into testing where you need it

File: test/suites/sharing/ram_share_test.go [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+		Eventually(func(g Gomega) {
+			// Policy status should be Accepted
+			gwNamespacedName := types.NamespacedName{
+				Name:      gateway.Name,
+				Namespace: gateway.Namespace,
+			}
+			gw := &gwv1.Gateway{}
+			err := testFramework.Client.Get(ctx, gwNamespacedName, gw)
+			g.Expect(err).To(BeNil())
+			g.Expect(len(gw.Status.Conditions)).To(BeEquivalentTo(2))
```
Comment (mikhail-aws) [inline test/suites/sharing/ram_share_test.go:158]: g.Expect(len(gw.Status.Conditions)).To(BeEquivalentTo(2))

not correct assertion, it should be at least 2. Also not related to RAM sharing test. Can be removed.

File: test/suites/sharing/ram_share_test.go [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+			g.Expect(gw.Status.Conditions[0].Type).To(BeEquivalentTo(string(gwv1.GatewayConditionAccepted)))
+			g.Expect(gw.Status.Conditions[0].Status).To(BeEquivalentTo(metav1.ConditionTrue))
+			g.Expect(gw.Status.Conditions[0].ObservedGeneration).To(BeEquivalentTo(1))
+			g.Expect(gw.Status.Conditions[0].Reason).To(BeEquivalentTo(string(gwv1.GatewayReasonAccepted)))
+			g.Expect(gw.Status.Conditions[1].Type).To(BeEquivalentTo(string(gwv1.GatewayConditionProgrammed)))
+			g.Expect(gw.Status.Conditions[1].Status).To(BeEquivalentTo(metav1.ConditionTrue))
+			g.Expect(gw.Status.Conditions[1].ObservedGeneration).To(BeEquivalentTo(1))
+			g.Expect(gw.Status.Conditions[1].Reason).To(BeEquivalentTo(string(gwv1.GatewayReasonProgrammed)))
```
Comment (mikhail-aws) [inline test/suites/sharing/ram_share_test.go:166]: Most of these are not related to ram share, checking for programmed status is enough.

File: test/suites/sharing/ram_share_test.go [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+			Expect(err).NotTo(HaveOccurred())
+
+			for _, inv := range listInvitationsResult.ResourceShareInvitations {
+				if *inv.ResourceShareName == randomName && *inv.SenderAccountId == secondaryAccountId {
+					invitation = inv
+					break
+				}
+			}
+			g.Expect(invitation).ToNot(BeNil())
+		}).Should(Succeed())
```
Comment (mikhail-aws) [inline test/suites/sharing/ram_share_test.go:206]: looks same as code in previous test, you can use same setup for these 2 tests

File: test/suites/sharing/ram_share_test.go [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+		}
+
+		deleteTaggedResources := func(
+			getResourcesResult *resourcegroupstaggingapi.GetResourcesOutput,
+			vpcLatticeClient *vpclattice.VPCLattice,
+			ramClient *ram.RAM,
+		) {
+			for _, mapping := range getResourcesResult.ResourceTagMappingList {
+				arn := mapping.ResourceARN
+				parts := strings.Split(*arn, ":")
```
Comment (mikhail-aws) [inline test/suites/sharing/ram_share_test.go:400]: use aws sdk arn.Parse()

File: test/suites/sharing/suite_test.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		associated, snva, _ := testFramework.IsVpcAssociatedWithServiceNetwork(ctx, vpcId, testServiceNetwork)
+		g.Expect(associated).To(BeTrue())
+		managed, _ := testFramework.Cloud.IsArnManaged(ctx, *snva.Arn)
+		g.Expect(managed).To(BeTrue())
+	}).Should(Succeed())
+
+}, func() {
+	testGateway = testFramework.NewGateway("test-gateway", k8snamespace)
+	testServiceNetwork = testFramework.GetServiceNetwork(ctx, testGateway)
+	testFramework.GrpcurlRunner = test.NewGrpcurlRunnerPod("grpc-runner", k8snamespace)
```
Comment (mikhail-aws) [inline test/suites/sharing/suite_test.go:76]: grpc curl for sharing tests?

**Review by mikhail-aws — CHANGES_REQUESTED** [review]
Status updates for gateway require more work and thought, existing code is not in good shape and new code does not make it better. I suggest move it into different PR.

Integ-tests have lots of duplicated code starting from Makefile, framework setup, and tests cases it-self. If you abstract same code into well defined functions, and re-use existing code as much as you can this PR would be under 200 lines of code.

**Comment by mikhail-aws** [comment]
Special highlight on duplicated code - last time when I did refactoring for FindServiceNetwork I had hard times with accesslog policy tests, surprisingly. We should take it more serious.

File: Makefile (hunk 2 above)
Comment (zijun726911) [reply Makefile:141]: What the specific reason we need a separate `./suites/sharing/` folder?

 Are we able to keep the new test file still in the `./suites/integration` and in the `developer.md` say:
to run the RAM share test, pass the FOCUS env variable? no?
```
FOCUS=RAM REGION=us-west-2 SECONDARY_ACCOUNT_TEST_ROLE_ARN=arn:aws:iam::000000000000:role/MyRole make sharing-test
```

Comment (xWink) [reply Makefile:141]: They require different environments due to needing the cluster VPC to have no service network association, plus the use of a secondary account. In the future, we may even want to test multi-cluster scenarios (2, maybe 3 accounts with their own clusters). Putting these tests in the old suite would be confusing, and if we are going to isolate them with FOCUS, then we may as well isolate them as separate suites.

File: Makefile (hunk 1 above)
Comment (xWink) [reply Makefile:127]: Will double check with only the above flags and remove the gingko ones if possible.

File: Makefile (hunk 2 above)
Comment (xWink) [reply Makefile:141]: Will look into skipping the RAM tests by default with ginkgo filtering specs. Then, if we pass a command to run the sharing tests, we run just that test.

File: pkg/model/core/status_condition.go (hunk 5 above)
Comment (xWink) [reply pkg/model/core/status_condition.go:5]: Good point, will move it there

File: pkg/utils/common.go (hunk 6 above)
Comment (xWink) [reply pkg/utils/common.go:96]: Will move to test code

File: pkg/config/controller_config.go (hunk 3 above)
Comment (xWink) [reply pkg/config/controller_config.go:26]: Will put this in Framework

File: test/suites/sharing/ram_share_test.go (hunk 7 above)
Comment (xWink) [reply test/suites/sharing/ram_share_test.go:158]: Good point, will do

File: test/suites/sharing/ram_share_test.go (hunk 8 above)
Comment (xWink) [reply test/suites/sharing/ram_share_test.go:166]: Will remove unrelated ones

File: test/suites/sharing/ram_share_test.go (hunk 10 above)
Comment (xWink) [reply test/suites/sharing/ram_share_test.go:400]: Thanks, will do

File: pkg/controllers/gateway_controller.go (hunk 4 above)
Comment (xWink) [reply pkg/controllers/gateway_controller.go:253]: Will change back to old error checking logic and add catch for name conflicts so that we can set the programmed status to Invalid instead of Conflicted

**Comment by xWink** [comment]
New E2E tests still pass:
```
Ran 3 of 61 Specs in 124.753 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 58 Skipped
--- PASS: TestIntegration (124.76s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/integration   126.001s

```

File: test/suites/integration/suite_test.go [hunk 12]
```diff
@@ -20,14 +23,30 @@ import (
 )
 
 const (
-	k8snamespace = "e2e-test"
+	k8snamespace                = "e2e-test"
+	secondaryAccountTestRoleArn = "SECONDARY_ACCOUNT_TEST_ROLE_ARN"
 )
 
+var k8sTestTags = map[string]*string{
+	anaws.TagBase + "TestSuite": aws.String("sharing"),
```
Comment (zijun726911) [inline test/suites/integration/suite_test.go:31]: Do you need to define `TestSuite:sharing` tag in the `ram_share_test.go` file? or change the tag value?

File: test/suites/integration/ram_share_test.go [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+		AllowExternalPrincipals: aws.Bool(true),
+		Tags:                    k8sRamTestTags,
+	}
+	_, err = secondaryRamClient.CreateResourceShareWithContext(ctx, createShareInput)
+	Expect(err).To(BeNil())
+
+	// Wait for resource share invitation to appear in primary account
+	var invitation *ram.ResourceShareInvitation = nil
+	Eventually(func(g Gomega) {
+		listInvitationsInput := &ram.GetResourceShareInvitationsInput{}
```
Comment (zijun726911) [inline test/suites/integration/ram_share_test.go:322]: Nit: I saw the vpc lattice canary python code did: `ram_client.get_resource_share_invitations(resourceShareArns=[resourceShareArn])`, Can you pass `resourceShareArns` in the `ram.GetResourceShareInvitationsInput`? and the response must have len(listInvitationsResult.ResourceShareInvitations)==1 or ==0 ?

**Review by zijun726911 — COMMENTED** [review]
Thanks to add new e2e tests! overall LGTM!

File: test/suites/integration/suite_test.go (hunk 12 above)
Comment (xWink) [reply test/suites/integration/suite_test.go:31]: Ah good point, this should just be the e2e test suite

File: test/suites/integration/ram_share_test.go (hunk 13 above)
Comment (xWink) [reply test/suites/integration/ram_share_test.go:322]: Thanks, will give this a shot

File: test/suites/integration/ram_share_test.go [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+		randomName := k8sResourceNamePrefix + utils.RandomAlphaString(10)
+
+		serviceNetwork := createSharedServiceNetwork(
+			randomName,
+			secondaryAccountId,
+			primaryVpcLatticeClient,
+			secondaryVpcLatticeClient,
+			primaryRamClient,
+			secondaryRamClient,
+		)
```
Comment (mikhail-aws) [inline test/suites/integration/ram_share_test.go:132]: since you are using same setup, you can create shared network once for all tests in BeforeAll

File: test/suites/integration/ram_share_test.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+})
+
+func createSharedServiceNetwork(
+	serviceNetworkName string,
+	secondaryAccountId string,
+	primaryVpcLatticeClient *vpclattice.VPCLattice,
+	secondaryVpcLatticeClient *vpclattice.VPCLattice,
+	primaryRamClient *ram.RAM,
+	secondaryRamClient *ram.RAM,
+) *vpclattice.GetServiceNetworkOutput {
```
Comment (mikhail-aws) [inline test/suites/integration/ram_share_test.go:299]: you can create struct that will hold all clients and create smaller methods that will create and delete resources

File: test/suites/integration/ram_share_test.go [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+			ResourceShareArns: []*string{createShareResult.ResourceShare.ResourceShareArn},
+		}
+		listInvitationsResult, err := primaryRamClient.GetResourceShareInvitations(listInvitationsInput)
+		Expect(err).NotTo(HaveOccurred())
+
+		if len(listInvitationsResult.ResourceShareInvitations) > 0 {
+			invitation = listInvitationsResult.ResourceShareInvitations[0]
+		}
+		g.Expect(invitation).ToNot(BeNil())
+	}).Should(Succeed())
```
Comment (mikhail-aws) [inline test/suites/integration/ram_share_test.go:332]: missing timeout and polling

**Review by mikhail-aws — CHANGES_REQUESTED** [review]

File: test/suites/integration/suite_test.go [hunk 17]
```diff
@@ -20,14 +23,30 @@ import (
...
 
+var k8sTestTags = map[string]*string{
+	anaws.TagBase + "TestSuite": aws.String(k8snamespace),
+}
+var k8sRamTestTags []*ram.Tag
 var testFramework *test.Framework
 var ctx context.Context
 var testGateway *gwv1.Gateway
 var testServiceNetwork *vpclattice.ServiceNetworkSummary
+var SecondaryAccountTestRoleArn string
```
[the remark this answers is not in the document]
Comment (mikhail-aws) [reply test/suites/integration/suite_test.go:38]: Also dont use global variables. It's extremely error-prone and painful to remove later on. Our config module is an example. We have Framework instance that holds all required variables.

File: test/suites/integration/ram_share_test.go [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+		primaryVpcLatticeClient       *vpclattice.VPCLattice
+		secondaryVpcLatticeClient     *vpclattice.VPCLattice
+		primaryResourceGroupsClient   *resourcegroupstaggingapi.ResourceGroupsTaggingAPI
+		secondaryResourceGroupsClient *resourcegroupstaggingapi.ResourceGroupsTaggingAPI
+	)
+
+	BeforeAll(func() {
+		parts := strings.Split(SecondaryAccountTestRoleArn, ":")
+		Expect(len(parts)).To(BeEquivalentTo(6), "Invalid secondary account role arn")
+		secondaryAccountId = parts[4]
```
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/ram_share_test.go:44]: Fixed next commit!

File: test/suites/integration/ram_share_test.go [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+			// Policy status should be Accepted
+			gwNamespacedName := types.NamespacedName{
+				Name:      gateway.Name,
+				Namespace: gateway.Namespace,
+			}
+			gw := &gwv1.Gateway{}
+			err := testFramework.Client.Get(ctx, gwNamespacedName, gw)
+			g.Expect(err).To(BeNil())
+			g.Expect(len(gw.Status.Conditions) >= 2).To(BeTrue())
+			g.Expect(gw.Status.Conditions[1].Type).To(BeEquivalentTo(string(gwv1.GatewayConditionProgrammed)))
```
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/ram_share_test.go:116]: Thanks for the tip! Fixed next commit!

File: test/suites/integration/ram_share_test.go (hunk 14 above)
Comment (xWink) [reply test/suites/integration/ram_share_test.go:132]: Name of service network depends specifically on the test. Will leave it as a function.

File: test/suites/integration/ram_share_test.go [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+	primaryRamClient *ram.RAM,
+	secondaryRamClient *ram.RAM,
+) *vpclattice.GetServiceNetworkOutput {
+	// Create secondary account's service network using randomName
+	createSNInput := &vpclattice.CreateServiceNetworkInput{
+		Name: aws.String(serviceNetworkName),
+		Tags: k8sTestTags,
+	}
+	createSNResult, err := secondaryVpcLatticeClient.CreateServiceNetworkWithContext(ctx, createSNInput)
+	Expect(err).NotTo(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/ram_share_test.go:306]: Will switch to using errors instead of Ginkgo functions

File: test/suites/integration/ram_share_test.go [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+	)
+
+	var (
+		secondaryAccountId            string
+		primaryRamClient              *ram.RAM
+		secondaryRamClient            *ram.RAM
+		primaryVpcLatticeClient       *vpclattice.VPCLattice
+		secondaryVpcLatticeClient     *vpclattice.VPCLattice
+		primaryResourceGroupsClient   *resourcegroupstaggingapi.ResourceGroupsTaggingAPI
+		secondaryResourceGroupsClient *resourcegroupstaggingapi.ResourceGroupsTaggingAPI
```
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/ram_share_test.go:38]: Good point. Shortened names next commit.

File: test/suites/integration/ram_share_test.go (hunk 15 above)
Comment (xWink) [reply test/suites/integration/ram_share_test.go:299]: Done next commit!

File: test/suites/integration/ram_share_test.go (hunk 16 above)
Comment (xWink) [reply test/suites/integration/ram_share_test.go:332]: Fixed next commit

File: test/suites/integration/suite_test.go [hunk 22]
```diff
@@ -20,14 +23,30 @@ import (
 )
 
 const (
-	k8snamespace = "e2e-test"
+	k8snamespace                = "e2e-test"
+	secondaryAccountTestRoleArn = "SECONDARY_ACCOUNT_TEST_ROLE_ARN"
```
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/suite_test.go]: Removed this line!

File: test/suites/integration/suite_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (xWink) [reply test/suites/integration/suite_test.go]: Done!

**Review by mikhail-aws — APPROVED** [review]
Nice, thank you

**Review by zijun726911 — APPROVED** [review]
